core_impl/src/main/java/io/opencensus/implcore/stats/MutableViewData.java

```diff
@@ -283,6 +289,213 @@ ViewData toViewData(Clock clock) {
+    private void shiftBucketList(long numOfPadBuckets, Timestamp now) {
+      Timestamp startOfNewBucket;
+      Duration totalDuration = Duration.fromMillis(N * toMillis(bucketDuration));
+
+      if (!buckets.isEmpty()) {
+        startOfNewBucket = buckets.peekLast().getStart().addDuration(bucketDuration);
+      } else {
+        // Initialize bucket list. Should only enter this block once.
+        startOfNewBucket = subtractDuration(now, totalDuration);
+      }
+
+      if (numOfPadBuckets > N + 1) {
+        // All current buckets expired, need to add N + 1 new buckets. The start time of the latest
+        // bucket will be current time.
+        startOfNewBucket = subtractDuration(now, totalDuration);
+        numOfPadBuckets = N + 1;
+      }
+
+      for (int i = 0; i < numOfPadBuckets; i++) {
+        buckets.add(
+            new IntervalBucket(startOfNewBucket, bucketDuration, super.view.getAggregations()));
+        startOfNewBucket = startOfNewBucket.addDuration(bucketDuration);
+      }
+
+      // removed expired buckets
+      while (buckets.size() > N + 1) {
+        buckets.pollFirst();
+      }
+    }
+
+    // Combine stats within each bucket, aggregate stats by tag values, and return the mapping from
+    // tag values to aggregation data.
+    private Map<List<TagValue>, List<AggregationData>> combineBucketsAndGetAggregationMap(
+        Timestamp now) {
+      Multimap<List<TagValue>, List<MutableAggregation>> multimap = HashMultimap.create();
+      LinkedList<IntervalBucket> shallowCopy = new LinkedList<IntervalBucket>(buckets);
+      List<Aggregation> aggregations = super.view.getAggregations();
+      putBucketsIntoMultiMap(shallowCopy, multimap, aggregations, now);
+      Map<List<TagValue>, List<MutableAggregation>> singleMap =
+          aggregateOnEachTagValueList(multimap, aggregations);
+      return createAggregationMap(singleMap);
+    }
+
+    // Put stats within each bucket to a multimap. Each tag value list (map key) could have multiple
+    // mutable aggregation lists (map value) from different buckets.
+    private static void putBucketsIntoMultiMap(
+        LinkedList<IntervalBucket> buckets,
+        Multimap<List<TagValue>, List<MutableAggregation>> multimap,
+        List<Aggregation> aggregations,
+        Timestamp now) {
+      // Put fractional stats of the head (oldest) bucket.
+      IntervalBucket head = buckets.peekFirst();
+      IntervalBucket tail = buckets.peekLast();
+      double fractionTail = tail.getFraction(now);
+      // TODO(songya): decide what to do when time goes backwards
+      checkArgument(0.0 <= fractionTail && fractionTail <= 1.0,
+          "Fraction " + fractionTail + " should be within [0.0, 1.0].");
+      double fractionHead = 1.0 - fractionTail;
+      putFractionalMutableAggregationsToMultiMap(
+          head.getTagValueAggregationMap(), multimap, aggregations, fractionHead);
+
+      // Put whole data of other buckets.
+      int i = 0;
```

Comment:
no need to count - consider making a bool (e.g. skip = true).

Reply:
Updated.